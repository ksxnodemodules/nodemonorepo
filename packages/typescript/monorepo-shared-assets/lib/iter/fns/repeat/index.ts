export type Iter<X> = IterableIterator<X>

export function * repeat<X> (iter: Iterable<X>, times: number): Iter<X> {
  if (times <= 0) return
  yield * iter
  yield * repeat(iter, times - 1)
}

export namespace repeat {
  export const element = <X>(x: X, times: number) => repeat([x], times)
}

export default repeat
