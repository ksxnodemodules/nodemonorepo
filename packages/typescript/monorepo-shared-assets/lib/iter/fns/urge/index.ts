import assert from 'assert'

export type Iter<X> = IterableIterator<X>

export function * urge<X> (
  iterable: Iterable<X>,
  partLength: number,
  handleRemain: urge.RemainingHandler<X> = urge.DEFAULT_REMAINING_HANDLER
): Iter<ReadonlyArray<X>> {
  assert(partLength > 0, `${partLength} > 0`)
  assert.equal(partLength % 1, 0, `${partLength} is an integer`)
  let count = partLength
  let tray = Array<X>()

  for (const x of iterable) {
    --count
    tray.push(x)

    if (!count) {
      yield tray
      count = partLength
      tray = []
    }
  }

  yield * handleRemain(tray)
}

export namespace urge {
  export const OMIT_EMPTY_REMAINING_PART = <X>(x: ReadonlyArray<X>) => x.length ? [x] : []
  export const KEEP_REMAINING_PART = <X>(x: X) => [x]
  export const OMIT_REMAINING_PART = () => []
  export const DEFAULT_REMAINING_HANDLER = OMIT_EMPTY_REMAINING_PART
  export type RemainingHandler<X> = (tray: ReadonlyArray<X>) => Iterable<ReadonlyArray<X>>
}

export default urge
