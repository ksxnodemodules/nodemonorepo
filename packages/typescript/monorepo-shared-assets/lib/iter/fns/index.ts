import range from './range'
import repeat from './repeat'
import urge from './urge'
export {range, repeat, urge}
