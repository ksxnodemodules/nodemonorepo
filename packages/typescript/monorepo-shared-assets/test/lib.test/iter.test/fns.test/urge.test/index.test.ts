import {iter} from '../../../../../index'
import fns = iter.fns
import range = fns.range
import urge = fns.urge

const PART_LENGTH = 4
const PART_COUNT = 5
const REMAIN_LENGTH = 3
const withRemaining = Array.from(range(PART_LENGTH * PART_COUNT + REMAIN_LENGTH))
const withoutRemaining = Array.from(range(PART_LENGTH * PART_COUNT))

const getArray = (
  iterable: Iterable<number>,
  fn?: urge.RemainingHandler<number>
) => Array.from(urge(iterable, 4, fn))

describe('omit only empty remaining part', () => {
  it('when some elements remain', () => {
    const array = getArray(withRemaining, urge.OMIT_EMPTY_REMAINING_PART)
    expect(array).toEqual(getArray(withRemaining)) // default parameter
    expect(array).toMatchSnapshot()
    expect(array.length).toBe(PART_COUNT + 1)
    expect(array[PART_COUNT].length).toBe(REMAIN_LENGTH)
  })

  it('when no elements remain', () => {
    const array = getArray(withoutRemaining, urge.OMIT_EMPTY_REMAINING_PART)
    expect(array).toEqual(getArray(withoutRemaining)) // default parameter
    expect(array).toMatchSnapshot()
    expect(array.length).toBe(PART_COUNT)
  })
})

describe('keep remaining part', () => {
  it('when some elements remain', () => {
    const array = getArray(withRemaining, urge.KEEP_REMAINING_PART)
    expect(array).toMatchSnapshot()
    expect(array.length).toBe(PART_COUNT + 1)
    expect(array[PART_COUNT].length).toBe(REMAIN_LENGTH)
  })

  it('when no elements remain', () => {
    const array = getArray(withoutRemaining, urge.KEEP_REMAINING_PART)
    expect(array).toMatchSnapshot()
    expect(array.length).toBe(PART_COUNT + 1)
    expect(array[PART_COUNT]).toEqual([])
  })
})

it('omit remaining part', () => {
  const array = getArray(withRemaining, urge.OMIT_REMAINING_PART)
  expect(array).toMatchSnapshot()
  expect(array.length).toBe(PART_COUNT)
  expect(array).toEqual(getArray(withoutRemaining, urge.OMIT_REMAINING_PART))
})
