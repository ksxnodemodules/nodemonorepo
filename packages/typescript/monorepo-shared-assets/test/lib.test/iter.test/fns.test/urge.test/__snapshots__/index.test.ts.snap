// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`keep remaining part when no elements remain 1`] = `
Array [
  Array [
    0,
    1,
    2,
    3,
  ],
  Array [
    4,
    5,
    6,
    7,
  ],
  Array [
    8,
    9,
    10,
    11,
  ],
  Array [
    12,
    13,
    14,
    15,
  ],
  Array [
    16,
    17,
    18,
    19,
  ],
  Array [],
]
`;

exports[`keep remaining part when some elements remain 1`] = `
Array [
  Array [
    0,
    1,
    2,
    3,
  ],
  Array [
    4,
    5,
    6,
    7,
  ],
  Array [
    8,
    9,
    10,
    11,
  ],
  Array [
    12,
    13,
    14,
    15,
  ],
  Array [
    16,
    17,
    18,
    19,
  ],
  Array [
    20,
    21,
    22,
  ],
]
`;

exports[`omit only empty remaining part when no elements remain 1`] = `
Array [
  Array [
    0,
    1,
    2,
    3,
  ],
  Array [
    4,
    5,
    6,
    7,
  ],
  Array [
    8,
    9,
    10,
    11,
  ],
  Array [
    12,
    13,
    14,
    15,
  ],
  Array [
    16,
    17,
    18,
    19,
  ],
]
`;

exports[`omit only empty remaining part when some elements remain 1`] = `
Array [
  Array [
    0,
    1,
    2,
    3,
  ],
  Array [
    4,
    5,
    6,
    7,
  ],
  Array [
    8,
    9,
    10,
    11,
  ],
  Array [
    12,
    13,
    14,
    15,
  ],
  Array [
    16,
    17,
    18,
    19,
  ],
  Array [
    20,
    21,
    22,
  ],
]
`;

exports[`omit remaining part 1`] = `
Array [
  Array [
    0,
    1,
    2,
    3,
  ],
  Array [
    4,
    5,
    6,
    7,
  ],
  Array [
    8,
    9,
    10,
    11,
  ],
  Array [
    12,
    13,
    14,
    15,
  ],
  Array [
    16,
    17,
    18,
    19,
  ],
]
`;
